signer.go

```diff
@@ -0,0 +1,188 @@
+		// x509.DSAWithSHA256:cryptoHash{algorithm: "dsa", hash: crypto.SHA256},
+		// Golang ECDSA support is lacking, can't seem to load private keys
+		// x509.ECDSAWithSHA1:   cryptoHash{algorithm: "ecdsa", hash: crypto.SHA1},
+		// x509.ECDSAWithSHA256: cryptoHash{algorithm: "ecdsa", hash: crypto.SHA256},
+		// x509.ECDSAWithSHA384: cryptoHash{algorithm: "ecdsa", hash: crypto.SHA384},
+		// x509.ECDSAWithSHA512: cryptoHash{algorithm: "ecdsa", hash: crypto.SHA512},
+	}
+}
+
+type cryptoHash struct {
+	algorithm string
+	hash      crypto.Hash
+}
+
+// Signer provides options for signing an XML document
+type Signer struct {
+	validator  *Validator
+	PrivateKey interface{}
+}
+
+// NewSigner returns a *Signer for the XML provided
+func NewSigner(xml string) (*Signer, error) {
+	v, err := NewValidator(xml)
+	if err != nil {
+		return nil, err
+	}
+	s := &Signer{validator: v}
+	return s, nil
+}
+
+// Sign populates the XML digest and signature based on the parameters present
+func (s *Signer) Sign() (string, error) {
+	if s.validator.signature == nil {
+		if err := s.validator.setEnvelopedSignature(); err != nil {
+			return "", err
+		}
+	}
+	if err := s.validator.setSignedInfo(); err != nil {
+		return "", err
+	}
+	if err := s.validator.setSigAlgorithm(); err != nil {
+		return "", err
+	}
+	if err := s.validator.setCanonAlgorithm(); err != nil {
+		return "", err
+	}
+	if err := s.setDigest(); err != nil {
+		return "", err
+	}
+	if err := s.setSignature(); err != nil {
+		return "", err
+	}
+
+	xml, err := s.validator.xml.WriteToString()
+	if err != nil {
+		return "", err
+	}
+	return xml, nil
+}
+
+func (s *Signer) setDigest() (err error) {
+	references := s.validator.signedInfo.FindElements("./Reference")
+	for _, ref := range references {
+		doc := s.validator.xml.Copy()
+		transforms := ref.SelectElement("Transforms")
+		for _, transform := range transforms.SelectElements("Transform") {
+			doc, err = processTransform(transform, doc)
+			if err != nil {
+				return err
+			}
+		}
+
+		doc, err := s.validator.getReferencedXML(ref, doc)
+		if err != nil {
+			return err
+		}
+
+		calculatedValue, err := calculateHash(ref, doc)
+		if err != nil {
+			return err
+		}
+
+		digestValueElement := ref.SelectElement("DigestValue")
+		if digestValueElement == nil {
+			return errors.New("signedxml: unable to find DigestValue")
+		}
+		digestValueElement.SetText(calculatedValue)
+	}
+	return nil
+}
+
+func (s *Signer) setSignature() error {
+	signedInfo, err := etree.CreateDocument(s.validator.signedInfo).WriteToString()
+	if err != nil {
+		return err
+	}
+
+	canonSignedInfo, err := s.validator.canonAlgorithm.Process(signedInfo, "")
+	if err != nil {
+		return err
+	}
+
+	var hashed, signature []byte
+	var h1, h2 *big.Int
+	signingAlgorithm, ok := signingAlgorithms[s.validator.sigAlgorithm]
+	if !ok {
+		return errors.New("signedxml: unsupported algorithm")
+	}
+
+	switch signingAlgorithm.hash {
```

Comment:
No need to use the switch block here.  Just call New() on signingAlgorithm.hash:
```go
hasher := signingAlgorithm.hash.New()
hasher.Write([]byte(canonSignedInfo))
hashed = hasher.Sum(nil)
```